src/main/java/io/nats/client/impl/Headers.java

```diff
@@ -575,19 +577,34 @@ boolean hasValues() {
 	 * @return the read only state
 	 */
 	public boolean isReadOnly() {
-		return readOnly;
+		return !(valuesMap instanceof HashMap);
 	}
 
 	@Override
 	public boolean equals(Object o) {
 		if (this == o) return true;
-		if (o == null || getClass() != o.getClass()) return false;
+		if (!(o instanceof Headers)) return false;
 		Headers headers = (Headers) o;
 		return Objects.equals(valuesMap, headers.valuesMap);
 	}
 
 	@Override
 	public int hashCode() {
-		return Objects.hash(valuesMap);
+		return Objects.hashCode(valuesMap);
+	}
+
+	@Override
+	public String toString() {
+		return valuesMap.entrySet().stream()
+				.filter(e -> e.getValue() != null)
+				.sorted(Map.Entry.comparingByKey())
+				.map(e -> {
+					String headerName = e.getKey();
+					List<String> values = e.getValue();
+					return headerName +": "+ (values.size() == 1 ? values.get(0)
+							: String.join(", ", values)
+					);
+				})
+				.collect(Collectors.joining("; "));
```

Comment:
how about just
```
return new String(getSerialized());
```

This has the benefit of preparing the serialization, so if no other values are added, it's ready to go across the wire, plus the toString exactly matches what goes over the wire. You could add a replace("\r\n", ";") if you don't want a multi line string.